src/plots/cartesian/index.js

```diff
@@ -25,61 +25,74 @@ exports.attrRegex = constants.attrRegex;
 
 exports.attributes = require('./attributes');
 
-exports.plot = function(gd) {
+exports.plot = function(gd, traces) {
+    var cdSubplot, cd, trace, i, j, k, isFullReplot;
+
     var fullLayout = gd._fullLayout,
         subplots = Plots.getSubplotIds(fullLayout, 'cartesian'),
         calcdata = gd.calcdata,
         modules = fullLayout._modules;
 
-    function getCdSubplot(calcdata, subplot) {
-        var cdSubplot = [];
-
-        for(var i = 0; i < calcdata.length; i++) {
-            var cd = calcdata[i];
-            var trace = cd[0].trace;
-
-            if(trace.xaxis + trace.yaxis === subplot) {
-                cdSubplot.push(cd);
-            }
+    if(!Array.isArray(traces)) {
+      // If traces is not provided, then it's a complete replot and missing
+      // traces are removed
+        isFullReplot = true;
+        traces = [];
+        for(i = 0; i < calcdata.length; i++) {
+            traces.push(i);
         }
-
-        return cdSubplot;
+    } else {
+      // If traces are explicitly specified, then it's a partial replot and
+      // traces are not removed.
+        isFullReplot = false;
     }
 
-    function getCdModule(cdSubplot, _module) {
-        var cdModule = [];
+    for(i = 0; i < subplots.length; i++) {
+        var subplot = subplots[i],
+            subplotInfo = fullLayout._plots[subplot];
+
+        // Get all calcdata for this subplot:
+        cdSubplot = [];
+        for(j = 0; j < calcdata.length; j++) {
+            cd = calcdata[j];
+            trace = cd[0].trace;
 
-        for(var i = 0; i < cdSubplot.length; i++) {
-            var cd = cdSubplot[i];
-            var trace = cd[0].trace;
+            // Skip trace if whitelist provided and it's not whitelisted:
+            // if (Array.isArray(traces) && traces.indexOf(i) === -1) continue;
 
-            if((trace._module === _module) && (trace.visible === true)) {
-                cdModule.push(cd);
+            if(trace.xaxis + trace.yaxis === subplot && traces.indexOf(trace.index) !== -1) {
+                cdSubplot.push(cd);
             }
         }
 
-        return cdModule;
-    }
-
-    for(var i = 0; i < subplots.length; i++) {
-        var subplot = subplots[i],
-            subplotInfo = fullLayout._plots[subplot],
-            cdSubplot = getCdSubplot(calcdata, subplot);
-
         // remove old traces, then redraw everything
-        // TODO: use enter/exit appropriately in the plot functions
-        // so we don't need this - should sometimes be a big speedup
-        if(subplotInfo.plot) subplotInfo.plot.selectAll('g.trace').remove();
+        // TODO: scatterlayer is manually excluded from this since it knows how
+        // to update instead of fully removing and redrawing every time. The
+        // remaining plot traces should also be able to do this. Once implemented,
+        // we won't need this - which should sometimes be a big speedup.
+        if(subplotInfo.plot) {
```

Comment:
Nice temporary workaround :+1: 

Could you add a comment on the line where class `trace` is added [here](https://github.com/plotly/plotly.js/pull/680/files/95f57a2751f8c8dae586ab42fcae76d6d2a15005#diff-ad4f76ccd6044ed16514297078e13b84R591) too?